### ISSUE TITLE
/usr/include/tbb/pipeline.h:328:74: error: 'has_trivial_copy_constructor' is not a member of 'std'

### ISSUE THREAD

**Issue by nschloe** [01org/tbb#12]
[A recent Debian test with GCC7](https://people.debian.org/~doko/logs/gcc7-20170126/trilinos_12.10.1-3_unstable_gcc7.log) failed with the TBB compilation error
```
/usr/include/tbb/pipeline.h:328:74: error: 'has_trivial_copy_constructor' is not a member of 'std'
```
(Coming from
https://github.com/01org/tbb/blob/a602c7dd54ca5e1d6b5a6b8e3289f0f70fc466a9/include/tbb/pipeline.h#L326.)

**Comment by turbo** [comment]
From the [GCC 7 WIP Docs](https://gcc.gnu.org/gcc-7/changes.html):

> The non-standard C++0x type traits has_trivial_default_constructor, has_trivial_copy_constructor and has_trivial_copy_assign have been removed.

This trait is not going to be part of GCC moving forward and is correctly marked as "not in std".

**Comment by daniel-j-h** [comment]
I just encountered the same issue with GCC 7.1 and a TBB release from 2015. Here's the code:

- https://github.com/01org/tbb/blob/2017_U7/include/tbb/pipeline.h#L324-L329
- https://github.com/01org/tbb/blob/2017_U7/include/tbb/tbb_config.h#L296-L297

GCC 7.1 works fine with the latest release [(2017 U7 atm)](https://github.com/01org/tbb/releases/tag/2017_U7) and probably all releases from 2017.

**Comment by akukanov** [comment]
The code in pipeline.h was fixed to use std::is_trivially_copyable when available as long as three years ago in TBB 4.3. The code still uses has_trivial_copy_constructor, but under preprocessor conditions that only allow its use with old compilers. Since then there were many other fixes and improvements for C++11 and newest compilers, so I highly recommend to update the version of TBB that you use.

**Comment by nschloe** [comment]
Thanks for the comment. I'll see what I can do to update Debian's TBB version.
